### ISSUE TITLE
[Question] APIService code structure and 3rd-parties networking library knowledge concerns

### ISSUE THREAD

**Issue by minhnimble** [nkhanh44/Survey_Nimble#28]
## Issue 

Looking at the current implementation for your `APIService` class, it is noticeable that for each of your `request` functions, the response handling contains major similarities with just some minor differences (i.e. `request` and `requestNoReply` functions have same response handling, just different decoder, `request` and `renewToken` functions have only differences in input type (generic vs user) and an additional `400 case`). Therefore, what is your plan to handle the refactor here to help avoid code cloning/duplication, adapt and retry the requests with more scalability, leading to effortless new extensions in the future like force-upgrade, force-logout, and SSL pinning down the road in each request here when requested?

https://github.com/nkhanh44/Survey_Nimble/blob/02fcb465d2a2c73e6bd62c817bc681c881d59e08/Survey-Nimble/Services/APIService.swift#L20-L282

Furthermore, I am curious if you are aware of or have the experience working with any 3rd-party networking library? As I notice that you are writing your `APIService` class purely with Apple's Foundation networking stack. What are your opinions on the pros and cons when handling networking using just Apple's Foundation networking? And when should we update to use 3rd-party libraries (e.g. Alamofire, Moya, etc.) for the application?

**Comment by nkhanh44** [comment]
> Therefore, what is your plan to handle the refactor here to help avoid code cloning/duplication, adapt and retry the requests with more scalability, leading to effortless new extensions in the future like force-upgrade, force-logout, and SSL pinning down the road in each request here when requested?

You're right ✌️ . Because of limited time and effort, i didn't care too much on the networking layer like the `APIService`, it maybe the least clean code in the project 😅 (but the rest is more clean :wink: ). 
So my refactoring plan is that decoupling `APIService's` into smaller parts to help avoid code cloning/duplication. I will work on the refactoring networking layer task tonight 😄 .

And for more scalability, I would like implement 3rd libraries (e.g Alamofire, Moya, APIKit etc). Because the framework is well maintained, tested, and used, also being familiar with most of iOS developers. 💯  

> Furthermore, I am curious if you are aware of or have the experience working with any 3rd-party networking library?

Yes ✌️ , every project I've joint using 3rd libraries (e.g Alamofire, Moya, APIKit) with different projects's size, structures, optimization and best practices.

> As I notice that you are writing your APIService class purely with Apple's Foundation networking stack. What are your opinions on the pros and cons when handling networking using just Apple's Foundation networking?

In my opinion, this project's size is kinda small so it's not necessary to bring a 3rd library to solve some basic API requests like GET, POST, refresh token ... etc. Therefore, I decided to choose building the networking layer from Apple's Foundation networking stack (URLSession), even though, i am not familiar with this one compared to 3rd libraries (e.g Alamofire, Moya, APIKit etc), especially using it with RxSwift.✌️

..And everything is a trade off. 

- **Pros** of this approach is that's suitable for a small project that handling only some basic cases. If we can use and understand them well enough, that must be a good foundation for us for building the scalable product with complex networking layer or using other 3rd party networking layer. :blush:

- **Cons** of this one is that it's too much lines of code for a request or set up, too much effort for maintenance, testing...etc

>  And when should we update to use 3rd-party libraries (e.g. Alamofire, Moya, etc.) for the application?

I think this application should be updated to use 3rd-party when there's upcoming features with many networking cases or when working in a team ✌️ . But again for this project scope, URLSession is fine enough in my opinion 😉 .

For sure, 3rd library like AlamoFire is a higher level of abstraction, more elegant with less lines of code, easy implement and usage, there are many available best practices for all most cases when dealing with networking as well. :sunglasses:

**Comment by nkhanh44** [comment]
Hello anh,
I refactored the APIService here ✌️ 
- https://github.com/nkhanh44/Survey_Nimble/pull/29
